### PR TITLE
Fixed #36064 -- Skipped an UPDATE when adding a model instance with a composite primary key with default values.

### PR DESCRIPTION
**Pull request by csirmazbendeguz**
#### Trac ticket number
ticket-36064

#### Branch description
If all fields of a CompositePrimaryKey has a default or db_default, force insert.

#### Checklist
- [X] This PR targets the `main` branch. 
- [X] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -1120,6 +1120,14 @@ def _save_table(
...
                 or (meta.pk.db_default and meta.pk.db_default is not NOT_PROVIDED)
+                or (
+                    meta.is_composite_pk
+                    and all(
+                        (f.default and f.default is not NOT_PROVIDED)
+                        or (f.db_default and f.db_default is not NOT_PROVIDED)
+                        for f in meta.pk
+                    )
+                )
             )
```
Comment (sarahboyce) [inline django/db/models/base.py:1131]: I might be wrong but I think we can simplify this slightly to:
```python
                meta.pk.has_default()
                or (meta.pk.db_default is not NOT_PROVIDED)
                or (
                    meta.is_composite_pk
                    and all(
                        f.has_default() or (f.db_default is not NOT_PROVIDED)
                        for f in meta.pk
                    )
                )
```
If so, we might want to spit to 2 commits with a refactor commit to have:
```python
                meta.pk.has_default()
                or (meta.pk.db_default is not NOT_PROVIDED)
```

Note: I think there might be some appetite to have a `has_db_default()` method in future

File: tests/composite_pk/test_create.py [hunk 2]
```diff
@@ -136,3 +137,12 @@ def test_update_or_create_user(self):
...
                 self.assertEqual(count + 1, User.objects.count())
+
+    def test_save_default_pk_not_set(self):
+        with self.assertNumQueries(1):
+            Post().save()
+
+    def test_save_default_pk_set(self):
+        post = Post.objects.create()
+        with self.assertRaises(IntegrityError):
+            Post(tenant_id=1, id=post.id).save()
```
Comment (sarahboyce) [inline tests/composite_pk/test_create.py:148]:
```suggestion
            Post(tenant_id=post.tenant_id, id=post.id).save()
```
(nit) I think this is slightly more explicit

**Review by sarahboyce — APPROVED** [review]
Thank you! I have minor comments
@LilyFoote can you also take a look?

File: django/db/models/base.py (hunk 1 above)
Comment (csirmazbendeguz) [reply django/db/models/base.py]: I was following the above lines so we'll need to change those as well I think

Comment (csirmazbendeguz) [reply django/db/models/base.py]: I don't know why would someone have such a pk default though, so maybe it's okay to change this.

Comment (sarahboyce) [reply django/db/models/base.py]: So looking into it this logic was added here 85458e94e38c20e57939947ee515a1a53689659f which added a test for when the primary key is a uuid and I feel like falsey default values for a primary key don't make much sense
But you're right that the logic isn't the same and perhaps someone relies on it, so let's not worry about it in this PR and this can be a discussion for another time :+1:

Comment (charettes) [reply django/db/models/base.py:1131]: @sarahboyce It seems off to me that we perform a boolean check against `meta.pk.default` and `meta.pk.db_default` since using `0` as primary key is also a valid value even if we put composite primary keys aside. Since primary keys cannot include `NULL` I think that checking against `is not None` makes more sense.

```suggestion
                (meta.pk.default is not None and meta.pk.default is not NOT_PROVIDED)
                or (meta.pk.db_default is not None and meta.pk.db_default is not NOT_PROVIDED)
                or (
                    meta.is_composite_pk
                    and all(
                        (f.default is not None and f.default is not NOT_PROVIDED)
                        or (f.db_default is not None and f.db_default is not NOT_PROVIDED)
                        for f in meta.pk
                    )
                )
            )
```

As to whether we should special case the composite primary key check I also get a sense that it might be more common so maybe we could start by using `is not None` there an open a follow up ticket to determine if we want to adjust the non-composite case as well?

Another approach that would simplify the whole thing (and avoid branching of `is_composite_pk`) is to simply use `Meta.pk_fields` systematically?

```suggestion
                all(
                    (f.default is not None and f.default is not NOT_PROVIDED)
                     or (f.db_default is not None and f.db_default is not NOT_PROVIDED)
                     for f in meta.pk_fields
                )
            )
```

Comment (csirmazbendeguz) [reply django/db/models/base.py:1131]: It's definitely more likely

The question is should we change this to force insert or leave it as it is?
I think yes. Whether the default is falsey shouldn't affect if force insert is triggered or not

Comment (csirmazbendeguz) [reply django/db/models/base.py:1131]: `f.default is not None` we could add a system check for this. We already enforce the composite pk fields are `null=False`, we can also enforce `default is not None`

Comment (charettes) [inline django/db/models/base.py:1131]: As brought up in the other comment it seems like both the falsey defaults and composite primary key support can be addressed by doing

```suggestion
            and all(
                (f.default is not None and f.default is not NOT_PROVIDED)
                or (f.db_default is not None and f.db_default is not NOT_PROVIDED)
                for f in meta.pk_fields
            )
```

Comment (jacobtylerwalls) [reply django/db/models/base.py:1131]: > @sarahboyce It seems off to me that we perform a boolean check against meta.pk.default and meta.pk.db_default since using 0 as primary key is also a valid value

Yep, I agree. The docs say "defines a default" which to me means `is not None`. I think the existing implementation (in the non-composite case) is wrong and the docs are right.

The other part of this logic also contravenes the docs: the docs mention if the pk is set to a truthy value (omitting 0 in the example, saying anything but "None or the empty string"), but we actually check it against None only. I think the implementation is correct here and we should change the docs. That's a separate to-be-filed ticket I think.

Back to the first issue, I doubt it really needs a deprecation path, as the original change setting this all in motion in 3.0 went straight into 3.0.

Comment (jacobtylerwalls) [reply django/db/models/base.py:1131]: > f.default is not None we could add a system check for this.

Ben, can I ask you to share why you think we should require defaults?

Comment (csirmazbendeguz) [reply django/db/models/base.py:1131]: I'm not saying require a default I'm saying add a check so `default` can't be set to `None`

Comment (sarahboyce) [reply django/db/models/base.py:1131]: Personally, I think combining Simon's suggestion and my original suggestion makes the most sense to me
```python
            and all(
                f.has_default() or f.db_default is not NOT_PROVIDED
                for f in meta.pk_fields
            )
``` 
I don't believe we have a test covering the original path of `default`/`db_default` Falsey but not `NOT_PROVIDED` cases
If we want to change to `default`/`db_default` is None, let's add a test for the None case.

Comment (csirmazbendeguz) [reply django/db/models/base.py:1131]: https://code.djangoproject.com/ticket/36076

Comment (csirmazbendeguz) [reply django/db/models/base.py]: Should I add a test for `default=False` or is this ok?

Comment (sarahboyce) [reply django/db/models/base.py]: I think I would like another test to be added as it adds more context to the commit around this discussion and would make something fail if this logic is reverted

File: django/db/models/base.py [hunk 3]
```diff
@@ -1117,9 +1117,9 @@ def _save_table(
             and not force_insert
             and not force_update
             and self._state.adding
-            and (
-                (meta.pk.default and meta.pk.default is not NOT_PROVIDED)
-                or (meta.pk.db_default and meta.pk.db_default is not NOT_PROVIDED)
+            and all(
+                f.has_default() or f.db_default is not NOT_PROVIDED
```
Comment (charettes) [inline django/db/models/base.py:1121]: Makes sense. FWIW I added the equivalent of `f.has_default()` but for `db_default` in [this PR](https://github.com/django/django/pull/18909/commits/e197a2b222696f246bc797d77b40267c702e8ae6) so we should eventually be able to get rid of this _ugly_ double negation and do

```suggestion
                f.has_default() or f.has_db_default()
```

**Review by charettes — APPROVED** [review]

**Review by LilyFoote — APPROVED** [review]

File: django/db/models/base.py (hunk 3 above)
Comment (charettes) [reply django/db/models/base.py:1121]: Created https://github.com/django/django/pull/19021 to include `has_db_default` if we want to get it in as well.

**Review by sarahboyce — APPROVED** [review]
Thank you all :star:
